### ISSUE TITLE
const problems with native handle

### ISSUE THREAD

**Issue by VictorEijkhout** [rabauke/mpl#29]
```
    const mpl::communicator &comm =
      mpl::environment::comm_world();
    MPI_Comm
      world_extract = comm.native_handle(),
```

gives `error: 'this' argument to member function 'native_handle' has type 'const mpl::communicator', but function is not marked const`

but without const

```
    mpl::communicator &comm =
      mpl::environment::comm_world();
```

gives `error: binding reference of type 'mpl::communicator' to value of type 'const mpl::communicator' drops 'const' qualifier`

**Comment by VictorEijkhout** [comment]
You could fix this by making `native_handle` a `const` function, but I gues the point to having the native handle is to do `MPI_Comm_set_info` and such on it.

**Comment by rabauke** [comment]
@VictorEijkhout: I was expecting this. You can replace `mpl::environment::comm_world()` just by `MPI_COMM_WOLD`. Thus, _this_ is not an issue.

Nevertheless, I will need to make the `native_handle` method `const` (not its return value). The reason is, that communicators should be passed as constant references to functions. (MPL communicators have a value semantics. Copying a communicator involves `MPI_Comm_dup`, which is usually undesirable.) Thus, if the underling MPI handle of the function argument is needed then we run into the same problem as above with `mpl::environment::comm_world()`.

**Comment by VictorEijkhout** [comment]
const mpl::communicator &comm =
      mpl::environment::comm_world();
    MPI_Comm
      world_extract = comm.native_handle(),
      world_given = MPI_COMM_WORLD;
    int result;
    MPI_Comm_compare(world_extract,world_given,&result);
    cout << "Compare raw comms: "
         << result << "\n”;

Should this give 0 or 1 ?

On , 2022Mar28, at 13:52, Heiko Bauke ***@***.******@***.***>> wrote:



Closed #29<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frabauke%2Fmpl%2Fissues%2F29&data=04%7C01%7C%7C145d5d61b6194ddd586708da10ec1374%7C31d7e2a5bdd8414e9e97bea998ebdfe1%7C0%7C0%7C637840903380027665%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=UTi%2B%2Bk39fTO%2FyDW6hD8eQc1nrhttL6%2BwDTSp6DjyYE4%3D&reserved=0> via ff9512f<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frabauke%2Fmpl%2Fcommit%2Fff9512fc61195b6c7e643e234789b0b937d28ee3&data=04%7C01%7C%7C145d5d61b6194ddd586708da10ec1374%7C31d7e2a5bdd8414e9e97bea998ebdfe1%7C0%7C0%7C637840903380027665%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=6BPLygjMSYDvsTxQTrKpIlvau%2Fw1VzEvIGFKgJFFYL0%3D&reserved=0>.

**Comment by VictorEijkhout** [comment]
My apologies. Everything works correctly.
